### PR TITLE
Generalize pit_loss to arbitrary shape and to retrieve the min perm

### PR DESCRIPTION
**Pull request by thequilo**
This PR generalizes `pit_loss` to use an arbitrary shape and axis.

It now uses python `min` instead of `torch.min` to get both the min loss and permutation at the same time. I am not sure if this is good or not; it makes the code easier to read but slower (I think).

### CODE REVIEW

**Comment by thequilo** [comment]
And I just noticed: I should add a doctest for more and less axis than 3

File: padertorch/ops/losses/loss.py [hunk 1]
```diff
@@ -105,12 +131,19 @@ def pit_loss(estimate, target, loss_fn=torch.nn.functional.mse_loss):
             f'{estimate.size()} != {target.size()}'
         )
     candidates = []
+    filler = (slice(None),) * axis
```
Comment (boeddeker) [inline padertorch/ops/losses/loss.py]: Did you mean `filler = (slice(None),) * estimate.size()`?

File: padertorch/ops/losses/loss.py [hunk 2]
```diff
@@ -105,12 +131,19 @@ def pit_loss(estimate, target, loss_fn=torch.nn.functional.mse_loss):
             f'{estimate.size()} != {target.size()}'
         )
     candidates = []
+    filler = (slice(None),) * axis 
     for permutation in itertools.permutations(range(sources)):
-        candidates.append(loss_fn(
-            estimate[:, permutation, :],
+        candidates.append((loss_fn(
+            estimate[filler + (permutation, )],
```
Comment (boeddeker) [inline padertorch/ops/losses/loss.py]: This looks wrong.
Wouldn't is be `filler[axis] = permutation` and `estimate[filler]`?

File: padertorch/ops/losses/loss.py (hunk 1 above)
Comment (thequilo) [reply padertorch/ops/losses/loss.py]: No, I meant to add `axis %= estimate.ndimension()`. But there are now to possibilites to solve this. My version: `estimate[filler + (permutation, )]`, your version would be `filler[axis] = permutation` and then index with `estimate[filler]`

File: padertorch/ops/losses/loss.py (hunk 2 above)
Comment (thequilo) [reply padertorch/ops/losses/loss.py]: Have a look at my previous comment, both are correct if `axis` is computed correctly

File: padertorch/ops/losses/loss.py [hunk 3]
```diff
@@ -105,12 +131,19 @@ def pit_loss(estimate, target, loss_fn=torch.nn.functional.mse_loss):
...
             target
-        ))
-    return torch.min(torch.stack(candidates))
+        ), permutation))
+
+    min_loss, min_perm = min(candidates, key=lambda x: x[0])
```
Comment (boeddeker) [inline padertorch/ops/losses/loss.py]: `torch.min` has the advantage that the execution is async on the GPU.
Python `min` introduces a sync point. It would be nice to have no sync point.

Comment (thequilo) [reply padertorch/ops/losses/loss.py]: It is not possible to have no sync point if we want to output the permutation (right?). We could use `torch.argmax` and then index into the list of permutations in case the permutation is requested. Then we would at least skip the sync point when the permutation is not requested

File: padertorch/ops/losses/loss.py (hunk 1 above)
Comment (boeddeker) [reply padertorch/ops/losses/loss.py]: Ah, you use another trick to get the axis position. Your solution is also fine.

File: padertorch/ops/losses/loss.py (hunk 3 above)
Comment (boeddeker) [reply padertorch/ops/losses/loss.py]: You are right. When the permutation is needed, we have to sync the GPU and the CPU.
Since in many cases we do not need the permutation, `torch.argmax` would be nicer, because then pytorch has the chance to do the operation without a sync.

**Comment by LukasDrude** [comment]
I suggest that `axis` is a mandatory argument. Otherwise `-2` seems to be a bit arbitrary default.

**Comment by thequilo** [comment]
This could potentially break code that uses the old `pit_loss` or `pit_mse_loss` functions. We could do `pit_mse_loss = partial(pit_loss, axis=-2)`.

**Comment by boeddeker** [comment]
I think a breaking change with such an error is ok.
In the doctest should then be a note that `axis=-2` will yield the old behavior.

**Comment by thequilo** [comment]
Should `axis` then be a mandatory argument or should it have a default value like `axis=0`?

**Comment by boeddeker** [comment]
I prefer to get the old behavior or the execution should fail.
Since `axis=0` may work or yield a strange error msg, I find `axis` as mandatory argument better.

**Comment by thequilo** [comment]
Same for `pit_mse_loss` or should this be held compatible with using `partial(pit_loss, axis=-2)`? `pit_mse_loss` is already marked with `# this function is kept at the moment for backwards compatibility` so I think we should keep compatibility here

**Comment by thequilo** [comment]
And how about `mse_loss` as default for `loss_fn`? This only makes sense for some PIT models, but there are many other cases where one might use different loss functions. I feel like from the name `pit_loss` the default argument `mse_loss` does not make too much sense.

**Comment by boeddeker** [comment]
I think MSE as default is fine. The first paper where PIT is present, they used MSE.
And MSE is a common loss function.

File: padertorch/ops/losses/loss.py [hunk 4]
```diff
@@ -105,16 +152,24 @@ def pit_loss(estimate, target, loss_fn=torch.nn.functional.mse_loss):
...
+
+    if return_permutation:
+        return min_loss, permutations[int(idx)]
+    else:
+        return min_loss
 
 # this function is kept at the moment for backwards compatibility
 # ToDo: remove this function
-pit_mse_loss = pit_loss
+pit_mse_loss = partial(pit_loss, axis=-2)
```
Comment (jensheit) [inline padertorch/ops/losses/loss.py]: I would suggest removing lines 169 to 172 and keep mse as default loss for pit

Comment (thequilo) [reply padertorch/ops/losses/loss.py]: `pit_mse_loss` is used in the `PermutationInvariantTraining` models. If we remove lines 169 to 172 we have to touch those models as well

Comment (thequilo) [reply padertorch/ops/losses/loss.py]: And some other people might use it in their personal repositories

Comment (jensheit) [reply padertorch/ops/losses/loss.py]: In that case I am fine with keeping it for the moment and removing it in a separate pull request together with all uses in padertorch.

**Comment by boeddeker** [comment]
LGTM. I will merge.
Deleting `pit_mse_loss` will be done in another PR.
